### PR TITLE
BYO Step

### PR DESCRIPTION
**Pull request by naterush**
# Description

This is a pretty cool PR, that does a few things:
1. Makes it so users can provide custom editors in Streamlit, Dash, and JupyterLab. These are just Python functions, and by requirement must take a dataframe as an input.
2. Makes it so users can pass dataframe and column header types, and we generate useful UIs for them. Right now, editors are required to have one dataframe, and the column headers reference the dataframe that most closely precedes them.

In doing so, I unified the custom importer and the custom edit function. This is pretty sweet, because now any improvements we make to the UI generation (of which there are many we could make to make these really polished / appealing!). 

The other major improvement here is that we now have a notion of a `runtime` action. This is something like a user defined importer or user defined editor. In turn, this means we can generate much better looking dropdown items -- one for each action -- and also have much better looking task panes.

# Testing

See the Streamlit testing file, and use some of those functions. Try creating your own editors and importers, as well.

# Documentation

We'll need to update the documentation for BYO importers, and update the documentation for BYO step. I am excited about this!

### CODE REVIEW

File: mitosheet/src/mito/components/toolbar/Toolbar.tsx [hunk 1]
```diff
@@ -78,14 +79,21 @@ export const Toolbar = (
...
 
-    importDropdownItems.push(
-        <DropdownItem title='Custom Imports' key='Custom Imports' onClick={() => {props.setUIState(prevUIState => {
-            return {
-                ...prevUIState,
-                currOpenTaskpane: {type: TaskpaneType.USERDEFINEDIMPORT}
-            }
-        })}}/>
-    )
+    props.actions.runtimeImportActionsList.map(action => {
```
Comment (naterush) [inline mitosheet/src/mito/components/toolbar/Toolbar.tsx:82]: Comment

File: mitosheet/src/mito/components/taskpanes/UserDefinedImport/UserDefinedFunctionParamConfigSection.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    if (paramNameToType === undefined || params === undefined) {
+        return <></>
+    }
+
+    const entries = Object.entries(paramNameToType);
+    console.log(entries);
+
+    const elements = []
+    let previousSheetIndex = -1;
+    for (let paramIndex = 0; paramIndex < entries.length; paramIndex++) {
```
Comment (naterush) [inline mitosheet/src/mito/components/taskpanes/UserDefinedImport/UserDefinedFunctionParamConfigSection.tsx:44]: Document all of this

Comment (naterush) [reply mitosheet/src/mito/components/taskpanes/UserDefinedImport/UserDefinedFunctionParamConfigSection.tsx:44]: And clean it up

File: mitosheet/mitosheet/step_performers/user_defined_edit.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    get_user_defined_function_param_type_and_execute_value_and_transpile_value
+from mitosheet.ai.recon import get_updated_state_from_reconed_dataframes
+from mitosheet.step_performers.utils.utils import get_param
+from mitosheet.types import ColumnID
+import pandas as pd
+
+
+class UserDefinedEditStepPerformer(StepPerformer):
+    """
+    Allows you to user defined edit.
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/user_defined_edit.py:28]: Update

**Review by naterush — COMMENTED** [review]
Cleanup!

File: mitosheet/mitosheet/tests/step_performers/test_user_defined_edit.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    
+def test_user_defined_edit_errors_with_wrong_first_param_type():
+    with pytest.raises(ValueError) as e_info:
+        create_mito_wrapper(editors=[wrong_first_param])
+
+def test_user_defined_edit_with_unqualified_column_header_errors():
+    with pytest.raises(ValueError) as e_info:
+        create_mito_wrapper(editors=[unqualified_column_header])
+
+def test_user_defined_edit_optimizes():
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/step_performers/test_user_defined_edit.py:177]: Can we add a test to make sure if you add a column to a dataframe, perform an edit to that dataframe and then delete the column, the code does not optimize out the add column. etc

**Comment by aarondr77** [comment]
I think it makes sense to show the types here so users know what to input. Having to hover over the icon to see that super helpful info is not that intuitive 

Screenshot 2023-10-15 at 11 14 45 AM

**Comment by aarondr77** [comment]
Can we make the inputs all the same width? 

Screenshot 2023-10-15 at 11 16 12 AM

**Comment by aarondr77** [comment]
The import name is overflowing the dropdown. We should update the dropdown to support

Screenshot 2023-10-15 at 11 16 39 AM

**Comment by aarondr77** [comment]
This error handling is a bit weird. If the user selects a column and then changes the dataframe, the selected column should probably reset to something valid. 

https://github.com/mito-ds/mito/assets/18709905/17afc49f-fbd3-4a4e-9a2d-2efb925ae26a

**Comment by aarondr77** [comment]
I think we should add a `custom edits` button the top row of the toolbar if the admin sets any. The only way I can find them now is by searching the Help dropdown which is pretty un discoverable. But I think in the case of the company we're building this for, the custom edit is going to be a prominent part of Mito usage and it should be easily discoverable. Also it creates more of a payoff for the admin to see it

**Review by aarondr77 — COMMENTED** [review]
This is a sweet feature! Definetly opens up more use cases that enterprises can support! 

I suggest a bit of polish below.

It's working pretty well from my initial testing. I will try to do some more testing today too, but want to get this review out.

File: mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+def validate_user_defined_editors(user_defined_editors: Optional[List[Callable]]) -> List[Callable]:
+    if user_defined_editors is None:
+        return []
+
+    for f in user_defined_editors:
+        df_arguments = []
+        parameters = inspect.signature(f).parameters
+        has_previous_dataframe_arg = False
+        for param_name in parameters:
+            annotation = parameters[param_name].annotation
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py:103]: We `param_type` elsewhere to refer to our type, so I actually think it disambiguates between the Python type and ours.

File: mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            elif param_type == 'str':
+                user_defined_function_params[param_name] = (param_type, param_value, get_column_header_as_transpiled_code(param_value))
+            elif param_type == 'int':
+                execute_value: Any = int(param_value)
+                user_defined_function_params[param_name] = (param_type, execute_value, get_column_header_as_transpiled_code(execute_value))
+            elif param_type == 'float':
+                execute_value = float(param_value) 
+                user_defined_function_params[param_name] = (param_type, execute_value, get_column_header_as_transpiled_code(execute_value))
+            elif param_type == 'bool':
+                execute_value = 'true' in param_value.lower()
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py:229]: We don't need to, as these are boolean toggle inputs.

File: mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                
+                # notably, the column header type will actually be provided as a column id? Does that make sense? I think so -- that's what we can 
+                # select on in the frontend. TODO: think about if this makes sense with the way replaying works...
+                column_header = state.column_ids.get_column_header_by_id(sheet_index, param_value)
+                user_defined_function_params[param_name] = (param_type, column_header, get_column_header_as_transpiled_code(param_value))
+            elif param_type == 'str':
+                user_defined_function_params[param_name] = (param_type, param_value, get_column_header_as_transpiled_code(param_value))
+            elif param_type == 'int':
+                execute_value: Any = int(param_value)
+                user_defined_function_params[param_name] = (param_type, execute_value, get_column_header_as_transpiled_code(execute_value))
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/step_performers/utils/user_defined_function_utils.py:224]: get_column_header_as_transpiled_code is just the name of the function that transpires a Python variable to a string - we renamed it recently. So it's not necessarily required for int and float, but I think it's clear and reasonable to keep.

**Comment by naterush** [comment]
LGTM when tests pass.

**Pull request merged**